Keyboard Cleanup + Event-Driven API

gyrovorbis opened the pull request:
Massive refactor of the keyboard Driver and API with a lot of new features. Everything that used to work still works and has been tested. The existing methods for processing input (ie: key press queue) are all there and work fine (and are now configurable). The ONLY thing that I did change is the names of the struct members in `kbd_state_t`, because I really feel like it had to be done to make the polling API pristine...

- Cleaned up C typing with typedefs, enums, unionized bitfields, and booleans
- Added an event-driven callback mechanism for registering a callback to be invoked whenever a key changes states (similar to how Qt, GTk, and high-level widgeting frameworks process key events)
- Made `kbd_state_t` a whole lot more user-friendly, as it's a public structure and is supposed to be reached into when polling for key states directly.
- Moved a bunch of crap that was cluttering the header and `kbd_state_t` from the public API to the private source file, where it's hidden from the user. This includes an "extended" `kbd_private_state_t` structure which contains all of the public members PLUS hidden private members only accessible from within the driver.
- `kbd_state_t`'s "key matrix" has been modified to store both the current frame and previous frame's key states, which allows for the driver to check for state changes and emit events accordingly. This added zero extra space overhead, since we were storing key states within a byte anyway, and I'm only using two bits of it.
- Added utility function for converting raw key to ASCII values without needing to go through a device
- Made repeated key press interval configurable 
- Adjusted libDream's keyboard example to show off event-driven handler mechanism
- Redid the entirety of the Doxygen with revamped topics, better organization, and example code snippets

andressbarajas commented on examples/dreamcast/libdream/keyboard/keyboard.c:
```diff
@@ -46,30 +42,36 @@ void kb_test(void) {
 
+static void on_key_event(maple_device_t *dev, key_state_t state,
+                         kbd_key_t key, kbd_mods_t mods,
+                         kbd_leds_t leds, void *user_data)
```
```suggestion
                         kbd_leds_t leds, void *user_data) {
```

gyrovorbis commented:
Looks like I might not be able to finish this tonight before passing out... A couple notes for me for when I pick this back u in a couple of days:

1) Somehow the regular keyboard example from Quzar is broken but the libDream keyboard example is working fine?
2) Need to add example code for the various ways to query input to the main keyboard Doxygen topic
3) Need to preview the Doxygen and make sure nothing looks screwed up.

QuzarDC commented on kernel/arch/dreamcast/hardware/maple/keyboard.c:
```diff
@@ -4,7 +4,11 @@
    Copyright (C) 2002 Megan Potter
    Copyright (C) 2012 Lawrence Sebald
    Copyright (C) 2018 Donald Haase
+<<<<<<< HEAD
```
tsk tsk

gyrovorbis replied:
OH NOEZ YOU GOT ME!!! I'm literally working on it right now and was like "WTF is this!?!"

Mefiresu commented on kernel/arch/dreamcast/include/dc/maple/keyboard.h:
```diff
@@ -28,283 +40,648 @@ __BEGIN_DECLS
+    device. This is what maple_dev_status() will return.
+*/
+typedef struct kbd_state {
+    /** \brief Current (and previous) state of all keys in kbd_keys_t. */
```
```suggestion
    /** \brief Current (and previous) state of all keys in kbd_keys_t */
```

Mefiresu commented on kernel/arch/dreamcast/hardware/maple/keyboard.c:
```diff
@@ -23,16 +24,88 @@ repeat handling.
+    repeat_timing.interval = interval;
+}
+
+/** \brief   Keyboard keymap.
```
```suggestion
/** \brief   Keyboard keymap
```

andressbarajas commented on kernel/arch/dreamcast/hardware/maple/keyboard.c:
```diff
@@ -23,16 +24,88 @@ repeat handling.
+    \ingroup kbd
+
+    This structure represents a mapping from raw key values to ASCII values, if
+    appropriate. This handles base values as well as shifted ("shift" and "Alt"
```
```suggestion
    appropriate. This handles base values as well as shifted ("Shift" and "Alt"
```

Mefiresu reviewed (commented):
Went through the Doxygen docs to look for typos and other minor formatting inconsistencies.

gyrovorbis commented:
Alrighty. I think it's time to take this bad-boy out of draft... Also here's some of the updated Doxygen documentation, after addressing review feedback:
![Screenshot from 2024-04-20 17-54-16](https://github.com/KallistiOS/KallistiOS/assets/5545520/8c90bb23-c399-4715-b81b-4a948586a08e)
![Screenshot from 2024-04-20 17-53-37](https://github.com/KallistiOS/KallistiOS/assets/5545520/3217813d-6e44-45dc-85fc-fb370b9fb825)

gyrovorbis commented:
Okay, I'm moving this back into draft until this PR: https://github.com/KallistiOS/KallistiOS/pull/517 gets merged. Which I will revise against, because I've totally gone to town on that area of the code.

gyrovorbis commented:
Okay, she's DONE. She's been done for awhile, but I had to make a teeeensy change to SDL to use the new API... because SDL was reaching inside of the state structure and using the raw `kbd_state_t::matrix` key state array, which had to be modified in order to support the new paradigm...

The old way really only stored the current frame's worth of up/down information. The new way and new state array uses the same amount of storage and stores two frames of information, so that you know when key up/down events have happened, which is what every engine has to implement one layer above this... (including my own and Simulant).

This is the corresponding SDL PR which uses the new state information: https://github.com/KallistiOS/kos-ports/pull/70.

gyrovorbis commented on doc/Doxyfile:
```diff
@@ -1013,7 +1013,7 @@ EXAMPLE_PATTERNS       =
 # irrespective of the value of the RECURSIVE tag.
 # The default value is: NO.
 
-EXAMPLE_RECURSIVE      = NO
```
The first one is for making it recursively search for examples when you provide it with an example directory and try to include some example code using `\include`. I never got it working or integrated with this PR, but we definitely don't want to have to include every example directory in this Doxygen config file when we do later.

The last one is to stop enums previews from blowing up the Doxygen when you have a massive enum declaration... They look horrible, and it's much cleaner and better looking to just have the enum name, brief description, then click on it to see the individual values rather than cramming them all into the index.

gyrovorbis replied:
The final addition tells Doxygen to ignore the `__packed` macro, because it tries to add it to the function declaration in the docs without resolving it, messing things up and looking wrong.

QuzarDC reviewed (commented):
I went through a good chunk. Most of the comments were *just* comments. Will try to get through the rest of it by the end of the week, but the breaking of the API is pretty deep. I was really surprised my 'keytest' didn't break, but at least it's good to know that it's not *mandatory* that every use would be broken. And I agree with the changes and that they're not really very avoidable to get the things done.

gyrovorbis commented on kernel/arch/dreamcast/include/dc/maple/keyboard.h:
```diff
@@ -28,283 +40,647 @@ __BEGIN_DECLS
+    \par Querying for a Device
+    You can grab a pointer to the `Nth` connected keyboard by using the following:
+
+        maple_device_t *device = maple_enum_type(N, MAPLE_FUNC_KEYBOARD);
+
+        if(device)
+            printf("Keyboard found!\n");
+        else
+            printf("Keyboard not found!\n");
```
Exactly! I think as we clean stuff up, we should be thinking about how to add little snippets like this to the high-level doxygen "groups" for how to use the API. Especially think this is warranted for the PVR API.

gyrovorbis commented on examples/dreamcast/libdream/keyboard/keyboard.c:
```diff
@@ -46,25 +40,40 @@ void kb_test(void) {
                 printf("Special key %04x\n", k);
 
-            if(k != 13) {
+            if(k != 10) {
```
Yeah, so with using `13` check out what happens after pressing return:
![Screenshot from 2024-06-05 15-50-25](https://github.com/KallistiOS/KallistiOS/assets/5545520/1a4358b1-20c2-4cb6-98c1-7a8c4b456fdb)
There's a blank character rendered in addition to the newline.

Then with the change, check it out, the next line starts flushed with the previous line:
![Screenshot from 2024-06-05 15-50-54](https://github.com/KallistiOS/KallistiOS/assets/5545520/f6721c5c-7aca-4e3a-b671-a6bb51b7fd92)

pcercuei replied:
Same here, don't use a magic value, use `'\r'` instead.

gyrovorbis commented on kernel/arch/dreamcast/include/dc/maple/keyboard.h:
```diff
@@ -28,283 +40,647 @@ __BEGIN_DECLS
+        2. `KEY_STATE_WAS_DOWN`: whether the key was previously pressed
+                                 last frame.
+
+    Between these two flags, you can know whether a key state transition
+    event has occurred (when the two flags have different values).
+
+    \sa key_state_t::raw, key_state_value_t
 
-    none-> pressed or none
-    was pressed-> pressed or none
-    pressed-> was_pressed
     @{
 */
-#define KEY_STATE_NONE        0
-#define KEY_STATE_WAS_PRESSED 1
-#define KEY_STATE_PRESSED     2
+#define KEY_STATE_IS_DOWN    (1 << 0) /**< \brief If key is currenty down */
+#define KEY_STATE_WAS_DOWN   (1 << 1) /**< \brief If key was previously down */
+/** \brief Mask of all key state flags */
+#define KEY_STATE_MASK      (KEY_STATE_IS_DOWN | KEY_STATE_WAS_DOWN)
```
Thanks, this is actually my second pass at a system like this, because I had one going in my original engine, then I took what I learned there and consulted what Simulant and Kazade were doing when constructing this. lol.

gyrovorbis commented on kernel/arch/dreamcast/include/dc/maple/keyboard.h:
```diff
@@ -28,283 +40,647 @@ __BEGIN_DECLS
+    This is a hardware constant. The define prevents the magic number '256'
+    from appearing.
 **/
-#define MAX_KBD_KEYS 256
```
It has to stay here in the header, because we're actually using it within the public segment of the structure for the key state array size at line `312`.

gyrovorbis commented on kernel/arch/dreamcast/include/dc/maple/keyboard.h:
```diff
@@ -28,283 +40,647 @@ __BEGIN_DECLS
 
-/** \brief   Keyboard raw condition structure.
-    \ingroup kbd
+/** \defgroup kbd_polling   State Polling
+    \brief                  Frame-based polling for keyboard input
+
+    One method of checking for key input is to simply poll
+    kbd_state_t::key_states for the desired key states each frame.
+
+    First, lets grab a pointer to the kbd_state_t:
```
EXACTLY. This is the only change that will have to be made to mode to the new model, and I made sure it was documented for people.

gyrovorbis commented on kernel/arch/dreamcast/include/dc/maple/keyboard.h:
```diff
@@ -28,283 +40,647 @@ __BEGIN_DECLS
+/** \defgroup   kbd_led_flags   Flags
+    \brief                      Keyboard LED flags
+
+    These are the LEDs that can be turned on and off on the keyboard. This list
+    may not be exhaustive. Think of these sort of like an extension of the
```
Yep, exactly. I went and consulted Flycast and asked around. Unfortunately like... I don't know that this stuff was ever actually used on a real keyboard, but it should still be here, cuz it's in maple and could be in the future potentially.

gyrovorbis commented on kernel/arch/dreamcast/hardware/maple/keyboard.c:
```diff
@@ -23,16 +26,88 @@ repeat handling.
+    event_handler.ud = user_data;
+}
+
 /* These are global timings for key repeat. It would be possible to put
     them in the state, but I don't see a reason to.
     It seems unreasonable that one might want different repeat
     timings set on each keyboard.
     The values are arbitrary based off a survey of common values. */
-uint16 kbd_repeat_start = 600, kbd_repeat_interval = 20;
+static struct {
+    uint16_t start;
+    uint16_t interval;
+} repeat_timing = {
+    600, 20
+};
```
Yeah, I really didn't see any reason to duplicate them over each keyboard state or convolute the API with making the user configure them all for each individual keyboard. It seems janky that an application would want different timeout and button repeat mechanics for each keyboard device, and it's saving RAM (well, it will in the future when maple devices are dynamically allocated)... Still felt like the way to go to me.

gyrovorbis commented on kernel/arch/dreamcast/hardware/maple/keyboard.c:
```diff
@@ -23,16 +26,88 @@ repeat handling.
+    NULL, NULL
+};
+
+void kbd_set_event_handler(kbd_event_handler_t callback, void *user_data) {
+    event_handler.cb = callback;
+    event_handler.ud = user_data;
+}
```
It's the latter. That the interior handler would delegate it out whatever needs it if they need that kind of behavior... 

...but now that I think about it, if they need that kind of behavior, they need a GETTER to save its state and restore it later......... good catch.........

gyrovorbis replied:
Added, btw.

gyrovorbis commented:
@QuzarDC I addressed all of your (actually fantastic) code review feedback. Thanks for being so thorough. I marked the ones as resolved that were obvious but left the others for you to resolve should you feel like I've adequately addressed your points. 

> I went through a good chunk. Most of the comments were _just_ comments. Will try to get through the rest of it by the end of the week, but the breaking of the API is pretty deep. I was really surprised my 'keytest' didn't break, but at least it's good to know that it's not _mandatory_ that every use would be broken. And I agree with the changes and that they're not really very avoidable to get the things done.

Actually, the only thing that should have broken that I know of is literally just anyone who had to reach into the `kbd_state_t` struct to directly access the key state via the `marix` array... And actually, none of the KOS examples even did that, and probably many people aren't doing that. The thing that broke was just SDL, cuz it had to do that to report back key press events.

Then the others who I knew for sure would be impacted were Kazade and SWAT. I contacted both of them to plead my case and beg for forgiveness, explaining what the change was, why I feel like it was justified, and told them what the teeensy little fix would be (essentially same as https://github.com/KallistiOS/KallistiOS/pull/517), and they both didn't hate me for it and were fine with it, thankfully... lol.

pold500 commented on examples/dreamcast/libdream/keyboard/keyboard.c:
```diff
@@ -1,9 +1,11 @@
 #include <kos.h>
 
-void kb_test(void) {
+static int x = 20, y = 20 + 24;
```
Why making this static? 
We can pass these variables as parameters to a function void kb_test(...), also, 20 and 24 can be named key values.
Like KB_KEY_D or something like that.

gyrovorbis replied:
Sorry for the late response. Getting back to KOS work now.

The values are really just for the starting x,y coordinates on the screen for where we're drawing text. So it starts out with margins of 20x20px... I think the 24 comes from the height of one font character?

At very least, these can be constants for the margins and then use the font height...

Knowing what these magic numbers represent, do you still think they should get passed in? I guess `main()` could pass them by pointer and be the one initializing and maintaining them across calls...

pcercuei replied:
Unresolving this conversation. We should not use magic values here. If 'x = 20' corresponds to the left margin, then it should be `LEFT_MARGIN`. If 'y = 20 + 24' corresponds to the top margin + the font size in pixels, then it should be `y = TOP_MARGIN + FONT_HEIGHT;`.

Also, don't make static variables with such short names! Inside the same file you actually do redefine 'x' and 'y' variables, that shadow the static ones, that's messy and asking for trouble.

Finally... I also see no reason why you use static values...

pold500 commented on examples/dreamcast/libdream/keyboard/keyboard.c:
```diff
@@ -28,16 +30,8 @@ void kb_test(void) {
         /* Get queued keys */
-        while((k = kbd_queue_pop(kbd, 1)) != -1) {
+        while((k = kbd_queue_pop(kbd, 1)) != KBD_QUEUE_END) {
             if(k == 27) {
```
Values like '27' should really be inside enum.
```
enum class EKeyboardKeys 
{
      KB_KEY_ESCAPE = 27
}
```

gyrovorbis replied:
There's an enum for all of the keyboard keys, but this API is special in that it's not popping off the actual keyboard key number, it's popping off the ASCII value the key corresponds to, hence 27... but yeah, probably at very least this could be a `#define` or an enum or even a comment... fair.

pcercuei replied:
27 is a magic number to me. Use `'\e'` instead, aka. the ASCII code for the ESC character.

gyrovorbis commented:
Alrighty, dudes, brought this bad boy back up-to-date, should be ready for mergery.

andressbarajas commented:
@QuzarDC Anything we can do to get this PR merged?

pcercuei commented on examples/dreamcast/libdream/keyboard/keyboard.c:
```diff
@@ -46,25 +40,40 @@ void kb_test(void) {
-                y += 24;
+                if(x >= 620) {
+                    x = 20;
+                    y += 24;
```
No magic values - should use LEFT_MARGIN / FONT_HEIGHT.

pcercuei commented on examples/dreamcast/libdream/keyboard/keyboard.c:
```diff
@@ -46,25 +40,40 @@ void kb_test(void) {
+
+    if(key == KBD_KEY_ENTER && state.value == KEY_STATE_CHANGED_DOWN) {
+        x = 20;
+        y += 24;
```
Ditto.

pcercuei commented on kernel/arch/dreamcast/include/dc/maple/keyboard.h:
```diff
@@ -28,283 +40,664 @@ __BEGIN_DECLS
+#define KBD_MOD_RCTRL       (1 << 4)    /**< \brief Right Control key */
+#define KBD_MOD_RSHIFT      (1 << 5)    /**< \brief Right Shift key */
+#define KBD_MOD_RALT        (1 << 6)    /**< \brief Right Alternate key */
+#define KBD_MOD_S2          (1 << 7)    /**< \brief S2 key */
```
Use the BIT() macro from `<kos/regfield.h>`.

pcercuei commented on kernel/arch/dreamcast/include/dc/maple/keyboard.h:
```diff
@@ -28,283 +40,664 @@ __BEGIN_DECLS
+#define KBD_LED_UNKNOWN2    (1 << 4)    /**< \brief Unknown LED 2 */
+#define KBD_LED_KANA        (1 << 5)    /**< \brief Kana LED */
+#define KBD_LED_POWER       (1 << 6)    /**< \brief Power LED */
+#define KBD_LED_SHIFT       (1 << 7)    /**< \brief Shift LED */
```
Use the BIT() macro from `<kos/regfield.h>`.

pcercuei commented on kernel/arch/dreamcast/include/dc/maple/keyboard.h:
```diff
@@ -28,283 +40,664 @@ __BEGIN_DECLS
-#define KBD_LED_CAPSLOCK    (1<<1)
-#define KBD_LED_SCRLOCK     (1<<2)
+#define KEY_STATE_IS_DOWN    (1 << 0) /**< \brief If key is currenty down */
+#define KEY_STATE_WAS_DOWN   (1 << 1) /**< \brief If key was previously down */
```
Use BIT() macro.

pcercuei commented on kernel/arch/dreamcast/include/dc/maple/keyboard.h:
```diff
@@ -28,283 +40,664 @@ __BEGIN_DECLS
+    /** \brief Key transitioned from down to released this frame */
+    KEY_STATE_CHANGED_UP   = KEY_STATE_PACK(false, true),
+    /** \brief Key has been held down for at least the last two frames */
+    KEY_STATE_HELD_DOWN    = KEY_STATE_PACK(true, true),
```
Sounds to me like the enum values need no explicit initialization.

pcercuei commented on examples/dreamcast/libdream/keyboard/keyboard.c:
```diff
@@ -46,25 +40,40 @@ void kb_test(void) {
-    for(y = 0; y < 480; y++)
-        for(x = 0; x < 640; x++) {
+    for(int y = 0; y < 480; y++)
+        for(int x = 0; x < 640; x++) {
```
As stated in my other comment, you're overshadowing static variables here (which had no reason to be static, and even less being one-character long).

pcercuei commented on include/kos/cdefs.h:
```diff
@@ -3,7 +3,7 @@
    Copyright (C) 2002, 2004 Megan Potter
    Copyright (C) 2020, 2023 Lawrence Sebald
-   Copyright (C) 2023 Falco Girgis
+   Copyright (C) 2023, 2024 Falco Girgis
```
I guess make it 2023-2025 now :)

pcercuei commented on include/kos/version.h:
```diff
@@ -99,7 +99,7 @@
 #define KOS_VERSION_MAJOR   2   /**< KOS's current major revision number. */
 #define KOS_VERSION_MINOR   1   /**< KOS's current minor revision number. */
-#define KOS_VERSION_PATCH   0   /**< KOS's current patch revision number. */
+#define KOS_VERSION_PATCH   1   /**< KOS's current patch revision number. */
```
That's not really up to a PR to update these fields. NAK.

pcercuei commented on kernel/arch/dreamcast/hardware/maple/keyboard.c:
```diff
@@ -5,11 +5,14 @@
    Copyright (C) 2012 Lawrence Sebald
    Copyright (C) 2018 Donald Haase
    Copyright (C) 2024 Paul Cercueil
+   Copyright (C) 2024 Falco Girgis
```
2024-2025

pcercuei commented on kernel/arch/dreamcast/hardware/maple/keyboard.c:
```diff
@@ -460,122 +537,154 @@ int kbd_get_key(void) {
+
+
+char kbd_key_to_ascii(kbd_key_t key, kbd_region_t region, kbd_mods_t mods, kbd_leds_t leds) {
+    char ascii = '\0';
```
Every code path override this variable, so it does not need to be initialized.

pcercuei commented on kernel/arch/dreamcast/hardware/maple/keyboard.c:
```diff
@@ -460,122 +537,154 @@ int kbd_get_key(void) {
-    else
-        ascii = keymaps[state->region - 1].base[(uint8)rv];
+    mods.raw = (rv >> 8) & 0xff;
+    leds.raw = (rv >> 16) & 0xff;
```
Use FIELD_GET() from `<kos/regfield.h>`.

pcercuei commented on kernel/arch/dreamcast/hardware/maple/keyboard.c:
```diff
@@ -460,122 +537,154 @@ int kbd_get_key(void) {
 
     /* Process all pressed keys */
-    for(i = 0; i < MAX_PRESSED_KEYS; i++) {
+    for(unsigned p = 0; p < KBD_MAX_PRESSED_KEYS; ++p) {
```
`unsigned` -> `unsigned int`

pcercuei commented on kernel/arch/dreamcast/hardware/maple/keyboard.c:
```diff
@@ -460,122 +537,154 @@ int kbd_get_key(void) {
+    const uint32_t mods = cond->modifiers.raw | (cond->leds.raw << 8);
+
+    /* Update all key states */
+    for(unsigned k = 0; k < KBD_MAX_KEYS; ++k) {
```
`unsigned` -> `unsigned int`

gyrovorbis commented:
I've reintroduced the corresponding SDL change here as well: https://github.com/KallistiOS/kos-ports/pull/100

gyrovorbis commented:
HOKAY BOIS.

I've addressed all of the review feedback (mostly on a test app I didn't write, lol)... couple little cleanups in the driver, such as for using the `<kos/regfield.h`> header.

Then I *went to town* on the example cleaning it up, properly documenting it, redoing a bunch of it... It's all must stronger now and is good to go!

QuzarDC commented:
Closing this one out as #984 finished getting its changes in.
